### PR TITLE
Adding sub status codes for 503's

### PR DESCRIPTION
**Pull request by aayush3011**
# Description

### Client SubStatusCodes

- TransportGenerated410 : TransportException creating a 410. Converted to 503 after retries in GoneAndRequestRetryPolicy.

- TimeoutGenerated410 : 410 generated due to timeout (ThrowGoneifElapsed). Converted to 503 after retries.

### Server SubStatus Codes

- ServerGenerated410 : Service generated 410 with 0 sub status code. Converted to 503 after retries

- GlobalStrongWriteBarrierNotMet: 410 created by ConsistencyWriter. Converted to 503 after retries

- ReadQuorumNotMet: 410 created QuorumReader. Converted to 503 after retries.

- ServerGenerated503: Service generated 503 with 0 sub status code

- NameCacheIsStaleExceededRetryLimit: 410 with namecachestaleexception. Converted to 503.

- PartitionKeyRangeGoneExceededRetryLimit: 410 with PartitionKeyRangeGoneException. Converted to 503.

- CompletingSplitExceededRetryLimit: 410 with PartitionKeyRangeIsSplittingException. Converted to 503.

- CompletingPartitionMigrationExceededRetryLimit: 410 with CompletingPartitionMigrationException. Converted to 503.

- NoValidStoreResponse: 410 thrown when there is no valid store response from the backend.

.Net PR 
- https://msdata.visualstudio.com/CosmosDB/_git/CosmosDB/pullrequest/710037?path=/Product/SDK/.net/Microsoft.Azure.Documents.Client/ClientRetryPolicy.cs&_a=files

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Review by FabianMeiswinkel — COMMENTED** [review]
Overall looks good to me - please get Annie to review as well and please add a markup document in the docs section describing all these sub status codes - + the other ones that existed before for at least 410 and 404:1002 - that way we can point customers and CSS at one place listing the meaning of sub status codes.

**Review by kushagraThapar — CHANGES_REQUESTED** [review]

**Review by kushagraThapar — CHANGES_REQUESTED** [review]
Can we have unit tests for these changes, so that we can catch any future regressions on these sub-status codes.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ConsistencyWriter.java [hunk 1]
```diff
@@ -288,7 +289,7 @@ Mono<StoreResponse> barrierForGlobalStrong(RxDocumentServiceRequest request, Sto
                 if (lsn.v == -1 || globalCommittedLsn.v == -1) {
                     logger.error("ConsistencyWriter: lsn {} or GlobalCommittedLsn {} is not set for global strong request",
                         lsn, globalCommittedLsn);
-                    throw new GoneException(RMResources.Gone);
+                    throw new GoneException(RMResources.Gone, HttpConstants.SubStatusCodes.SERVER_GENERATED_410);
```
[the remark this answers is not in the document]
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ConsistencyWriter.java:293]: Discussed Offline

**Review by kushagraThapar — APPROVED** [review]
LGTM, thanks @aayush3011

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/HttpConstants.java [hunk 2]
```diff
@@ -412,8 +412,27 @@ public static class SubStatusCodes {
...
-
+      
         public static final int USER_REQUEST_RATE_TOO_LARGE = 3200;
+
+        //SDK Codes(Client)
+        public static final int TRANSPORT_GENERATED_410 = 20001;
+        public static final int TIMEOUT_GENERATED_410 = 20002;
+        public static final int TRANSPORT_GENERATED_503 = 20003;
+        public static final int CPU_OVERLOAD = 20004;
+        public static final int THREAD_STARVATION = 20005;
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/HttpConstants.java:423]: I do not think we check THREAD_STARVATION in java

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/ServiceUnavailableException.java [hunk 3]
```diff
@@ -79,5 +86,12 @@ public ServiceUnavailableException(String message,
             HttpUtils.asMap(headers),
             HttpConstants.StatusCodes.SERVICE_UNAVAILABLE,
             requestUriString);
+        setSubStatus(subStatusCode);
+    }
+
+    private void setSubStatus(int subStatusCode) {
+        this.getResponseHeaders().put(
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/ServiceUnavailableException.java:93]: I guess we can also use `BridgeInternal.setSubStatusCode(this, subStatusCode);`

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/QuorumReader.java [hunk 4]
```diff
@@ -202,7 +203,7 @@ public Mono<StoreResponse> readStrongAsync(
                             case QuorumNotSelected:
                                 if (hasPerformedReadFromPrimary.v) {
                                     logger.warn("QuorumNotSelected: Primary read already attempted. Quorum could not be selected after retrying on secondaries.");
-                                    return Flux.error(new GoneException(RMResources.ReadQuorumNotMet));
+                                    return Flux.error(new GoneException(RMResources.ReadQuorumNotMet, HttpConstants.SubStatusCodes.READ_QUORUM_NOT_MET));
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/QuorumReader.java:206]: I kind feel we should differentiate READ_QUORUM_NOT_MET vs READ_QUORUM_NOT_SELECTED. (but non blocking, just bring up the question, and also understand this is .Net behavior)

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java [hunk 5]
```diff
@@ -927,7 +928,7 @@ private void completeAllPendingRequestsExceptionally(
             final Map<String, String> requestHeaders = record.args().serviceRequest().getHeaders();
             final String requestUri = record.args().physicalAddressUri().getURI().toString();
 
-            final GoneException error = new GoneException(message, cause, null, requestUri);
+            final GoneException error = new GoneException(message, cause, null, requestUri, HttpConstants.SubStatusCodes.UNKNOWN);
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java:931]: should this be Transport_Generated_410?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java [hunk 6]
```diff
@@ -1055,7 +1056,8 @@ private void messageReceived(final ChannelHandlerContext context, final RntbdRes
                             break;
                         default:
                             GoneException goneExceptionFromService =
-                                new GoneException(error, lsn, partitionKeyRangeId, responseHeaders);
+                                new GoneException(error, lsn, partitionKeyRangeId, responseHeaders,
+                                    HttpConstants.SubStatusCodes.UNKNOWN);
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java:1060]: should this be Server_Generated_410?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java [hunk 7]
```diff
@@ -1088,15 +1090,17 @@ private void messageReceived(final ChannelHandlerContext context, final RntbdRes
 
                 case StatusCodes.REQUEST_TIMEOUT:
                     Exception inner = new RequestTimeoutException(error, lsn, partitionKeyRangeId, responseHeaders);
-                    cause = new GoneException(resourceAddress, error, lsn, partitionKeyRangeId, responseHeaders, inner);
+                    cause = new GoneException(resourceAddress, error, lsn, partitionKeyRangeId, responseHeaders, inner,
+                        HttpConstants.SubStatusCodes.UNKNOWN);
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java:1094]: I wonder whether we should use a different substatusCode for this case as well

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java [hunk 8]
```diff
@@ -1088,15 +1090,17 @@ private void messageReceived(final ChannelHandlerContext context, final RntbdRes
...
                     break;
 
                 case StatusCodes.RETRY_WITH:
                     cause = new RetryWithException(error, lsn, partitionKeyRangeId, responseHeaders);
                     break;
 
                 case StatusCodes.SERVICE_UNAVAILABLE:
-                    cause = new ServiceUnavailableException(error, lsn, partitionKeyRangeId, responseHeaders);
+                    cause = new ServiceUnavailableException(error, lsn, partitionKeyRangeId, responseHeaders,
+                        HttpConstants.SubStatusCodes.UNKNOWN);
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java:1103]: should this be server_service_unavailable or similar?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/BridgeInternal.java [hunk 9]
```diff
@@ -653,7 +654,8 @@ public static ConsistencyLevel fromServiceSerializedFormat(String consistencyLev
 
     @Warning(value = INTERNAL_USE_ONLY_WARNING)
     public static CosmosException createServiceUnavailableException(Exception innerException) {
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/BridgeInternal.java:656]: this method is being called by GoneAndRetryPolicy -> without passing in the substatuscode, the correct substatus code will still be getting lost.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/HttpConstants.java (hunk 2 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/HttpConstants.java:423]: please remove `CPU_OVERLOAD`, `THREAD_STARVATION` for now as they are not being used as of today.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/HttpConstants.java [hunk 10]
```diff
@@ -412,8 +412,27 @@ public static class SubStatusCodes {
...
+
+        //SDK Codes (Server)
+        public static final int NAME_CACHE_IS_STALE_EXCEEDED_RETRY_LIMIT = 21001;
+        public static final int PARTITION_KEY_RANGE_GONE_EXCEEDED_RETRY_LIMIT = 21002;
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/HttpConstants.java:427]: The following status code is not being used:
```

  public static final int NAME_CACHE_IS_STALE_EXCEEDED_RETRY_LIMIT = 21001;
        public static final int PARTITION_KEY_RANGE_GONE_EXCEEDED_RETRY_LIMIT = 21002;
        public static final int COMPLETING_SPLIT_EXCEEDED_RETRY_LIMIT = 21003;
        public static final int COMPLETING_PARTITION_MIGRATION_EXCEEDED_RETRY_LIMIT = 21004;
```

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/AddressSelector.java [hunk 11]
```diff
@@ -61,7 +62,7 @@ public static Uri getPrimaryUri(RxDocumentServiceRequest request, List<AddressIn
             // Primary endpoint (of the desired protocol) was not found.
             throw new GoneException(String.format("The requested resource is no longer available at the server. Returned addresses are {%s}",
                                                   String.join(",", replicaAddresses.stream()
-                                                      .map(address -> address.getPhysicalUri().getURIAsString()).collect(Collectors.toList()))), null);
+                                                      .map(address -> address.getPhysicalUri().getURIAsString()).collect(Collectors.toList()))), null, HttpConstants.SubStatusCodes.UNKNOWN);
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/AddressSelector.java:65]: nit: there is a method public GoneException(String msg) {} which will automatically populate UNKNOWN status code, you do not need to explicitly populate it

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/FailFastRntbdRequestRecord.java [hunk 12]
```diff
@@ -58,7 +58,8 @@ public static FailFastRntbdRequestRecord createAndFailFast(
         final GoneException admissionControlBlocksRequestException = new GoneException(
             reason,
             headers,
-            remoteAddress);
+            remoteAddress,
+            HttpConstants.SubStatusCodes.UNKNOWN);
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/FailFastRntbdRequestRecord.java:62]: Let us use TRANSPORT_GENERATED_410 for now

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java [hunk 13]
```diff
@@ -927,7 +928,7 @@ private void completeAllPendingRequestsExceptionally(
             final Map<String, String> requestHeaders = record.args().serviceRequest().getHeaders();
             final String requestUri = record.args().physicalAddressUri().getURI().toString();
 
-            final GoneException error = new GoneException(message, cause, null, requestUri);
+            final GoneException error = new GoneException(message, cause, null, requestUri, SubStatusCodes.UNKNOWN);
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java:931]: use TRANSPORT_GENERATED_410

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java [hunk 14]
```diff
@@ -1088,15 +1090,17 @@ private void messageReceived(final ChannelHandlerContext context, final RntbdRes
 
                 case StatusCodes.REQUEST_TIMEOUT:
                     Exception inner = new RequestTimeoutException(error, lsn, partitionKeyRangeId, responseHeaders);
-                    cause = new GoneException(resourceAddress, error, lsn, partitionKeyRangeId, responseHeaders, inner);
+                    cause = new GoneException(resourceAddress, error, lsn, partitionKeyRangeId, responseHeaders, inner,
+                        SubStatusCodes.UNKNOWN);
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java:1094]: use SERVER_GENERATED_408

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestRecord.java [hunk 15]
```diff
@@ -260,7 +261,7 @@ public boolean expire() {
             // Convert from requestTimeoutException to GoneException for the following two scenarios so they can be safely retried:
             // 1. RequestOnly request
             // 2. Write request but not sent yet
-            error = new GoneException(this.toString(), null, this.args.physicalAddressUri().getURI());
+            error = new GoneException(this.toString(), null, this.args.physicalAddressUri().getURI(), HttpConstants.SubStatusCodes.UNKNOWN);
```
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestRecord.java:264]: use TRANSPORT_GENERATED_410

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/ServiceUnavailableException.java (hunk 3 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/ServiceUnavailableException.java:93]: Discussed offline, going to have a followup PR to refactor it as there are multiple duplicate places

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/QuorumReader.java (hunk 4 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/QuorumReader.java:206]: Discussed offline, align with .Net behavior for now.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java (hunk 6 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java]: updated.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java (hunk 8 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java]: Updated

**Review by xinlian12 — APPROVED** [review]
Thanks @aayush3011 for the change

- There are few cases missing, please check where we should populate them, can do in a following PR:

`CPU_OVERLOAD`
`THREAD_STARVATION`
`NAME_CACHE_IS_STALE_EXCEEDED_RETRY_LIMIT`
`PARTITION_KEY_RANGE_GONE_EXCEEDED_RETRY_LIMIT`
`COMPLETING_SPLIT_EXCEEDED_RETRY_LIMIT`
`COMPLETING_PARTITION_MIGRATION_EXCEEDED_RETRY_LIMIT`

- There are few places, please populate Transport_generated_410
- Can you also share few diagnostics which would show the new status code

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/BridgeInternal.java (hunk 9 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/BridgeInternal.java:656]: Fixed the code.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/HttpConstants.java (hunk 10 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/HttpConstants.java:427]: Getting used now in a new method ```getExceptionSubStatusCodeForGoneRetryPolicy```
which returns the subStatusCode to be added in the ```createServiceUnavailableException``` in the ``` GoneAndRetryWithRetryPolicy``` class

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/AddressSelector.java (hunk 11 above)
Comment (aayush3011) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/AddressSelector.java:65]: Fixed the code

**Review by xinlian12 — APPROVED** [review]
LGTM, thanks

**Pull request merged**